torch/lib/c10d/TCPStore.cpp

```diff
@@ -176,14 +176,16 @@ void TCPStoreDaemon::compareSetHandler(int socket) {
   std::vector<uint8_t> currentValue = tcputil::recvVector<uint8_t>(socket);
   std::vector<uint8_t> newValue = tcputil::recvVector<uint8_t>(socket);
 
-  // sets value to new value if the key exists and it's existing value is the
-  // current value
-  auto it = tcpStore_.find(key);
-  if (it != tcpStore_.end() && it->second == currentValue) {
-    it->second = newValue;
-    tcputil::sendVector<uint8_t>(socket, newValue);
-  } else {
+  auto pos = tcpStore_.find(key);
+  if (pos == tcpStore_.end()) {
+    // TODO: This code path is not ideal as we are "lying" to the caller in case
```

Comment:
So now we can return 3 possible values? For example, if the existing key/value pair is "key"/"value" and the method is called with compare_set("key", "current_value", "new_value") it will return "value"

I think I saw in the standalone rendezvous RFC requirements for TCPStore that we would just return 2 possible values, "current_value" and "new_value," unless I misunderstood the use case.

Reply:
Yes, you are right. The possible actions that `compare_set` can take are:
1. Return the current value if the key exists and the passed `current_value` is different than the real current value.
2. Return the new value if the key exists and the passed `current_value` is identical to the real current value.
3. **_Return `None` or raise a `KeyError` if the key is not in the store._**

Right now the implementation does not handle the 3rd point and as a temporary solution just returns the passed `current_value` as you already mentioned. And this is why I added a TODO to that code path. This "hack" unblocks other tasks that depend on TCPStore, but as soon as we finish those, we should fix this.

Reply:
Sounds good! Thanks for clarifying